File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution(object):
    def is_possible(self,weights,mid,days):
        day=1
        ship_load=0
        i=0
        while(i< len(weights)):
            if(ship_load + weights[i] <= mid):
                ship_load += weights[i]
                i +=1
            else:
                day +=1
                ship_load =0
            
            if(day > days): 
                return False

        return True

    def shipWithinDays(self, weights, days):
        """
        :type weights: List[int]
        :type days: int
        :rtype: int
        """
        start = 1
        total =0 
        for i in weights:
            total += i
        end = total
        ans=0
        while(start <= end):
            mid = (start+end) // 2

            if(self.is_possible(weights,mid,days)):
                ans = mid
                end = mid-1
            else:
                start = mid +1
        return ans
        

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
    bool is_possible(vector<int> &weights , int mid , int days){
        int day=1;
        int ship_load =0;

        for(int i=0 ; i< weights.size() ;i++){
            if(weights[i]+ship_load <= mid){
                ship_load += weights[i];
            }
            else{
                day+=1;
                ship_load =weights[i];
            }

            if(day > days){
                return false;
            }
        }
        return true;
    }
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int total=0;
        int maxi = weights[0];

        for(int i =0 ; i< weights.size() ; i++){
            total += weights[i];
            maxi = max(maxi , weights[i]);
        }

        int start = maxi;
        int end = total;
        int ans =-1;

        while(start <= end){
            int mid= (start+end) /2;

            if(is_possible(weights,mid,days)){
                ans = mid;
                end = mid-1;
            }
            else{
                start = mid+1;
            }
        }
        return ans;
    }
};